core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/upscale/StackUpscaleService.java

```diff
@@ -0,0 +1,243 @@
+        Set<Resource> retSet = new HashSet<>();
+        for (CloudResourceStatus cloudResourceStatus : cloudResourceStatuses) {
+            if (!cloudResourceStatus.isFailed()) {
+                CloudResource cloudResource = cloudResourceStatus.getCloudResource();
+                Resource resource = new Resource(cloudResource.getType(), cloudResource.getName(), cloudResource.getReference(), cloudResource.getStatus(),
+                        stack, null);
+                retSet.add(resource);
+            }
+        }
+        return retSet;
+    }
+
+    private void validateResourceResults(CloudContext cloudContext, Exception exception, List<CloudResourceStatus> results) {
+        String action = "upscale";
+        if (exception != null) {
+            LOGGER.error(format("Failed to %s stack: %s", action, cloudContext), exception);
+            throw new OperationException(exception);
+        }
+        if (results.size() == 1 && (results.get(0).isFailed() || results.get(0).isDeleted())) {
+            throw new OperationException(format("Failed to %s the stack for %s due to: %s", action, cloudContext, results.get(0).getStatusReason()));
+        }
+    }
+
+    private void updateNodeCount(long stackId, List<Group> originalGroups, List<CloudResourceStatus> statuses, boolean create) {
+        for (Group group : originalGroups) {
+            int nodeCount = group.getInstances().size();
+            List<CloudResourceStatus> failedResources = removeFailedMetadata(stackId, statuses, group);
+            if (!failedResources.isEmpty() && create) {
+                int failedCount = failedResources.size();
+                InstanceGroup instanceGroup = instanceGroupRepository.findOneByGroupNameInStack(stackId, group.getName());
+                instanceGroup.setNodeCount(nodeCount - failedCount);
+                instanceGroupRepository.save(instanceGroup);
+                flowMessageService.fireEventAndLog(stackId, Msg.STACK_INFRASTRUCTURE_ROLLBACK_MESSAGE, Status.UPDATE_IN_PROGRESS.name(),
+                        failedCount, group.getName(), failedResources.get(0).getStatusReason());
+            }
+        }
+    }
+
+    private List<CloudResourceStatus> removeFailedMetadata(long stackId, List<CloudResourceStatus> statuses, Group group) {
+        Map<Long, CloudResourceStatus> failedResources = new HashMap<>();
+        Set<Long> groupPrivateIds = getPrivateIds(group);
+        for (CloudResourceStatus status : statuses) {
+            Long privateId = status.getPrivateId();
+            if (privateId != null && status.isFailed() && !failedResources.containsKey(privateId) && groupPrivateIds.contains(privateId)) {
+                failedResources.put(privateId, status);
+                instanceMetadataService.deleteInstanceRequest(stackId, privateId);
+            }
+        }
+        return new ArrayList<>(failedResources.values());
```

Comment:
Wrap isn't neccessary, if you replace .get(0) to .iterator().next(), and i guess it generates less resources.